narwhal/primary/src/certificate_fetcher.rs

```diff
@@ -474,7 +533,108 @@ async fn process_certificates_helper(
             .inc_by(now.elapsed().as_micros() as u64);
     }
 
-    trace!("Fetched certificates have been processed");
+    Ok(())
+}
+
+async fn process_certificates_v2_helper(
+    protocol_config: &ProtocolConfig,
+    response: FetchCertificatesResponse,
+    synchronizer: &Synchronizer,
+    metrics: Arc<PrimaryMetrics>,
+) -> DagResult<()> {
+    let mut all_certificates = vec![];
+    let mut all_parents = HashSet::new();
+    for cert in response.certificates {
+        // We should not be getting mixed versions of certificates from a
+        // validator, so any individual certificate with mismatched versions
+        // should cancel processing for the entire batch of fetched certificates.
+        all_certificates.push(
+            validate_received_certificate_version(cert.clone(), protocol_config).map_err(
+                |err| {
+                    error!("{err}");
+                    DagError::Canceled
+                },
+            )?,
+        );
+
+        for parent_digest in cert.header().parents() {
+            all_parents.insert(*parent_digest);
+        }
+    }
+
+    let all_certificates_count = all_certificates.len() as u64;
+
+    // Identify leaf certs and preemptively set the parent certificates
+    // as verified indirectly. This is safe because any leaf certs that
+    // fail verification will cancel processing for all fetched certs.
+    let mut leaf_certs = Vec::new();
+    for (idx, c) in all_certificates.iter_mut().enumerate() {
+        if !all_parents.contains(&c.digest()) {
```

Comment:
This is not enough to make sure all certificates have parents available in store or among other fetched certificates. e.g. a fetch certificate can exist in all_parents, but itself is missing some parents.

Reply:
But that is all checked in `try_accept_fetched_certificate` so it is okay, right? Only thing we are skipping is the signature verification.